include/VecRng/VecRNG.h

```diff
@@ -0,0 +1,215 @@
+#define VECRNG_VECRNG_H 1
+
+/**
+ * VecRNG: The base class of SIMD/SIMT random number generators
+ *
+ * Requirements :
+ * 1) DerivedT  : A pseudo-random number generator with multiple streams
+ * 2) BackendT  : Scalar, Vector, Cuda
+ * 3) RandomT   : A templated struct of DerivedT states with BackendT 
+ */
+
+#include "RngDefs.h"
+
+namespace vecRng {
+inline namespace VECRNG_IMPL_NAMESPACE {
+
+template <typename DerivedT, typename BackendT, typename RandomT>
+class VecRNG {
+
+protected:
+  // Use *this to access data members in the derived class
+  RandomT *fState;
+
+public:
+
+  VECCORE_ATT_HOST_DEVICE
+#ifndef VECCORE_CUDA
+  VecRNG() { fState = (RandomT *)AlignedAlloc(VECCORE_SIMD_ALIGN, sizeof(RandomT)); }
+#else
+  VecRNG() { fState = new RandomT; }
+#endif
+
+  // Dummy Constructor for SIMT
+  VECCORE_ATT_HOST_DEVICE
+  VecRNG(RandomT *devState) {}
+
+  VECCORE_ATT_HOST_DEVICE
+#ifndef VECCORE_CUDA
+  ~VecRNG() { AlignedFree(fState); }
+#else
+  ~VecRNG() { delete fState; }
+#endif
+
+  VECCORE_ATT_HOST_DEVICE
+  VecRNG(const VecRNG &rng) = default;
+
+  // Static interfaces (Required methods)
+  
+  // Initialization for SIMD
+  VECCORE_ATT_HOST 
+  void Initialize() 
+  { static_cast<DerivedT *>(this)->template Initialize<BackendT>(); }
+
+  // Initialization for SIMT
+  VECCORE_ATT_HOST 
+  void Initialize(RandomT *states, int blocks, int threads)
+  { static_cast<DerivedT *>(this)->template Initialize<BackendT>(states,blocks,threads); }
+
+  // Return ReturnTypeBackendT::Double_v of random numbers in [0,1)
+  template <typename ReturnTypeBackendT>
+  VECCORE_ATT_HOST_DEVICE 
+  typename ReturnTypeBackendT::Double_v Uniform() 
+  { return static_cast<DerivedT *>(this)->template Kernel<ReturnTypeBackendT>(*this->fState); }
+
+  // Generate random numbers based on a given state
+  template <typename ReturnTypeBackendT>
+  VECCORE_ATT_HOST_DEVICE 
+  typename ReturnTypeBackendT::Double_v Uniform(RandomT *state) 
+  { return static_cast<DerivedT *>(this)->template Kernel<ReturnTypeBackendT>(*state); }
+
+  // Auxiliary methods 
+
+  VECCORE_ATT_HOST_DEVICE 
+  void SetState(RandomT *state) { fState = state; }
+
+  VECCORE_ATT_HOST_DEVICE 
+  RandomT* GetState() const { return fState; }
+
+  VECCORE_ATT_HOST 
+  void PrintState() const { static_cast<DerivedT *>(this)->PrintState(); }
+
+  //Common methods
+
+  // Returns an array of random numbers of the type ReturnTypeBackendT::Double_v
+  template <typename ReturnTypeBackendT>
+  VECCORE_ATT_HOST_DEVICE 
+  void Array(const size_t nsize, typename ReturnTypeBackendT::Double_v *array);
+
+  // Flat distribution in [min,max)
+  template <typename ReturnTypeBackendT>
+  VECCORE_ATT_HOST_DEVICE 
+  typename ReturnTypeBackendT::Double_v Flat(typename ReturnTypeBackendT::Double_v min, 
+                                             typename ReturnTypeBackendT::Double_v max) 
+  { return min+(max-min)*static_cast<DerivedT *>(this)-> template Uniform<ReturnTypeBackendT>(); }
```

Comment:
It would be nice if the generator followed the interfaces of the standard library, so that it could be used in places where a standard generator is required. Only a few interfaces are required, so I believe it's worth adding them:
http://en.cppreference.com/w/cpp/concept/RandomNumberEngine

Reply:
Good point.  We will consider this when we discuss a common design at this/next weeks
with John and Lorenzo.